File: csharp-runtime/Envelope.cs
using System.Text;

namespace RosettaRpc;

/// <summary>
/// Wire-level envelope wrapping every response that crosses the RpcClient
/// dispatch boundary, so that errors can be propagated to the client.
///
/// Every language runtime implements the same envelope protocol:
/// 
/// <code>
/// envelope := status_byte ++ payload
/// status_byte == 0x00 (Ok)  -> payload = the encoded Resp protobuf, verbatim
/// status_byte == 0x01 (Err) -> payload = code:i32 (4 bytes LE)
///                                        ++ message_len:u32 (4 bytes LE)
///                                        ++ message (UTF-8 bytes)
/// </code>
/// </summary>
internal static class Envelope
{
    private const byte StatusOk = 0x00;
    private const byte StatusErr = 0x01;

    public static byte[] EncodeOk(byte[] payload)
    {
        var bytes = new byte[1 + payload.Length];
        bytes[0] = StatusOk;
        Array.Copy(payload, 0, bytes, 1, payload.Length);
        return bytes;
    }

    public static byte[] EncodeError(RpcErrorCode code, string message)
    {
        var messageBytes = Encoding.UTF8.GetBytes(message);
        var bytes = new byte[1 + 4 + 4 + messageBytes.Length];
        bytes[0] = StatusErr;
        WriteInt32LE((int)code, bytes, 1);
        WriteUInt32LE((uint)messageBytes.Length, bytes, 5);
        Array.Copy(messageBytes, 0, bytes, 9, messageBytes.Length);
        return bytes;
    }

    /// <summary>
    /// Decodes an envelope, returning the inner payload on success, or
    /// throwing <see cref="RpcException"/> on the error case (or on a
    /// malformed envelope, as <see cref="RpcErrorCode.Decode"/>).
    /// </summary>
    public static byte[] Decode(byte[] bytes)
    {
        if (bytes.Length < 1)
        {
            throw new RpcException(RpcErrorCode.Decode, "empty response envelope");
        }

        var status = bytes[0];
        var rest = bytes.AsSpan(1);
        return status switch
        {
            StatusOk => rest.ToArray(),
            StatusErr => HandleError(rest),
            _ => throw new RpcException(RpcErrorCode.Decode, $"unknown envelope status byte: {status}"),
        };
    }

    private static byte[] HandleError(ReadOnlySpan<byte> payload)
    {
        if (payload.Length < 8)
        {
            throw new RpcException(RpcErrorCode.Decode, "truncated error envelope");
        }
        var code = (RpcErrorCode)ReadInt32LE(payload);
        var messageLength = ReadUInt32LE(payload[4..]);
        var messageBytes = payload[8..];
        if (messageBytes.Length != messageLength)
        {
            throw new RpcException(RpcErrorCode.Decode, "truncated error envelope message");
        }
        var message = Encoding.UTF8.GetString(messageBytes);
        throw new RpcException(code, message);
    }

    private static void WriteInt32LE(int value, byte[] destination, int offset) =>
        WriteUInt32LE(unchecked((uint)value), destination, offset);

    private static void WriteUInt32LE(uint value, byte[] destination, int offset)
    {
        destination[offset] = (byte)value;
        destination[offset + 1] = (byte)(value >> 8);
        destination[offset + 2] = (byte)(value >> 16);
        destination[offset + 3] = (byte)(value >> 24);
    }

    private static int ReadInt32LE(ReadOnlySpan<byte> source) => unchecked((int)ReadUInt32LE(source));

    private static uint ReadUInt32LE(ReadOnlySpan<byte> source) =>
        (uint)source[0] | ((uint)source[1] << 8) | ((uint)source[2] << 16) | ((uint)source[3] << 24);
}

File: kotlin-runtime/src/main/kotlin/rosettarpc/Envelope.kt
package rosettarpc

import java.nio.charset.StandardCharsets

/**
 * Wire-level envelope wrapping every response that crosses the [RpcClient]/[ServiceRegistry]
 * dispatch boundary, so a call failure can travel back to the caller as data instead of relying on
 * an exception surviving Godot's `Variant` call boundary between language runtimes.
 *
 * Deliberately hand-rolled rather than a protobuf message: this framing is purely internal to this
 * library, never part of a project's `.proto` schema. The Rust and C# runtimes implement the same
 * framing — keep them in sync if this changes.
 *
 * ```
 * envelope := status_byte ++ payload
 * status_byte == 0x00 (Ok)  -> payload = the encoded Resp protobuf, verbatim
 * status_byte == 0x01 (Err) -> payload = code:i32 (4 bytes LE)
 *                                        ++ message_len:u32 (4 bytes LE)
 *                                        ++ message (UTF-8 bytes)
 * ```
 */
internal object Envelope {
  private const val STATUS_OK: Byte = 0x00
  private const val STATUS_ERR: Byte = 0x01

  fun encodeOk(payload: ByteArray): ByteArray = byteArrayOf(STATUS_OK) + payload

  fun encodeError(code: RpcErrorCode, message: String): ByteArray {
    val messageBytes = message.toByteArray(StandardCharsets.UTF_8)
    val bytes = ByteArray(1 + 4 + 4 + messageBytes.size)
    bytes[0] = STATUS_ERR
    writeInt32LE(code.wireValue, bytes, 1)
    writeInt32LE(messageBytes.size, bytes, 5)
    messageBytes.copyInto(bytes, destinationOffset = 9)
    return bytes
  }

  /**
   * Decodes an envelope, returning the inner payload on success, or throwing [RpcException] on the
   * error case (or on a malformed envelope, as [RpcErrorCode.DECODE]).
   */
  fun decode(bytes: ByteArray): ByteArray {
    if (bytes.isEmpty()) {
      throw RpcException(RpcErrorCode.DECODE, "empty response envelope")
    }
    val status = bytes[0]
    val rest = bytes.copyOfRange(1, bytes.size)
    return when (status) {
      STATUS_OK -> rest
      STATUS_ERR -> handleError(rest)
      else -> throw RpcException(RpcErrorCode.DECODE, "unknown envelope status byte: $status")
    }
  }

  private fun handleError(payload: ByteArray): ByteArray {
    if (payload.size < 8) {
      throw RpcException(RpcErrorCode.DECODE, "truncated error envelope")
    }
    val code = RpcErrorCode.fromWireValue(readInt32LE(payload, 0))
    val messageLength = readInt32LE(payload, 4)
    val messageBytes = payload.copyOfRange(8, payload.size)
    if (messageBytes.size != messageLength) {
      throw RpcException(RpcErrorCode.DECODE, "truncated error envelope message")
    }
    val message = String(messageBytes, StandardCharsets.UTF_8)
    throw RpcException(code, message)
  }

  private fun writeInt32LE(value: Int, destination: ByteArray, offset: Int) {
    destination[offset] = (value and 0xFF).toByte()
    destination[offset + 1] = ((value shr 8) and 0xFF).toByte()
    destination[offset + 2] = ((value shr 16) and 0xFF).toByte()
    destination[offset + 3] = ((value shr 24) and 0xFF).toByte()
  }

  private fun readInt32LE(source: ByteArray, offset: Int): Int =
      (source[offset].toInt() and 0xFF) or
          ((source[offset + 1].toInt() and 0xFF) shl 8) or
          ((source[offset + 2].toInt() and 0xFF) shl 16) or
          ((source[offset + 3].toInt() and 0xFF) shl 24)
}
